src/Integration.Vsix/CFamily/MessageHandler.cs

```diff
@@ -0,0 +1,97 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2021 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Diagnostics;
+using System.Linq;
+using SonarLint.VisualStudio.Core.Analysis;
+using SonarLint.VisualStudio.Core.CFamily;
+using SonarLint.VisualStudio.Core.Helpers;
+
+namespace SonarLint.VisualStudio.Integration.Vsix.CFamily
+{
+    /// <summary>
+    /// Handles messages returned by the CFamily subprocess.exe
+    /// </summary>
+    internal interface IMessageHandler
+    {
+        int IssueCount { get;  }
+
+        void HandleMessage(Message message);
+    }
+
+    /// <summary>
+    /// No-op implementation - used when there is not a valid issue consumer
+    /// </summary>
+    internal class NoOpMessageHandler : IMessageHandler
+    {
+        /// <summary>
+        /// Singleton no-op message handler
+        /// </summary>
+        public static readonly IMessageHandler Instance = new NoOpMessageHandler();
+
+        public int IssueCount { get; } = 0;
+        public void HandleMessage(Message message) { /* no-op */ }
+    }
+
+    internal class MessageHandler : IMessageHandler
+    {
+        private readonly IRequest request;
+        private readonly IIssueConsumer issueConsumer;
+        private readonly ICFamilyIssueToAnalysisIssueConverter issueConverter;
+
+        public int IssueCount { get; private set; }
+
+        public MessageHandler(IRequest request, IIssueConsumer issueConsumer, ICFamilyIssueToAnalysisIssueConverter issueConverter)
+        {
+            this.request = request;
+            this.issueConsumer = issueConsumer;
+            this.issueConverter = issueConverter;
+        }
+
+        public void HandleMessage(Message message)
+        {
+            Debug.Assert(PathHelper.IsMatchingPath(message.Filename, request.Context.File), $"Issue for unexpected file returned: {message.Filename}");
```

Comment:
In the next PR I'll move the filtering-on-filename from the `Protocol.Read` to this class.

I'll then start working on the new functionality i.e. processing the internal messages.